src/libopensc/card-eoi.c

```diff
@@ -0,0 +1,532 @@
+		goto err;
+	if (!EVP_EncryptInit(ctx, EVP_aes_256_ecb(), key, NULL))
+		goto err;
+	if (!EVP_EncryptUpdate(ctx, (u8 *)output, &len, input, AES_BLOCK_SIZE))
+		goto err;
+	/* Since it's exactly 1 block, no need to call EVP_EncryptFinal */
+	r = 1;
+err:
+	if (ctx)
+		EVP_CIPHER_CTX_free(ctx);
+	return r;
+}
+
+static int aes256_ecb_decrypt(const u8 *key, const u8 *input, const u8 *output)
+{
+	EVP_CIPHER_CTX *ctx = EVP_CIPHER_CTX_new();
+	int r = 0, len = AES_BLOCK_SIZE;
+	if (!ctx)
+		goto err;
+	if (!EVP_DecryptInit(ctx, EVP_aes_256_ecb(), key, NULL))
+		goto err;
+	/* Disable padding, otherwise it will fail to decrypt non-padded inputs */
+	if (!EVP_CIPHER_CTX_set_padding(ctx, 0))
+		goto err;
+	if (!EVP_DecryptUpdate(ctx, (u8 *)output, &len, input, len))
+		goto err;
+	/* Since it's exactly 1 block, no need to call EVP_DecryptFinal */
+	r = 1;
+err:
+	if (ctx)
+		EVP_CIPHER_CTX_free(ctx);
+	return r;
+}
+
+/*
+ * CAN decrypt magic...
+ */
+static int get_can_key(const u8 *key, const u8 round, const u8 *input, u8 *output)
+{
+	size_t i;
+	u8 tmp[3][AES_BLOCK_SIZE];
+	memset(tmp[0], 0, AES_BLOCK_SIZE);
+	if (!aes256_ecb_encrypt(key, tmp[0], tmp[0]))
+		return 0;
+	rol(tmp[1], tmp[0]);
+	rol(tmp[0], tmp[1]);
+	memset(tmp[1], 0, AES_BLOCK_SIZE);
+	tmp[1][11] = 4;
+	tmp[1][13] = 1;
+	tmp[1][15] = round;
+	if (!aes256_ecb_encrypt(key, tmp[1], tmp[2]))
+		return 0;
+	memset(tmp[1], 0, AES_BLOCK_SIZE);
+	memcpy(tmp[1], &input[AES_BLOCK_SIZE], 8);
+	tmp[1][8] = 0x80;
+	for (i = 0; i < AES_BLOCK_SIZE; i++)
+		tmp[0][i] = tmp[0][i] ^ tmp[1][i] ^ tmp[2][i];
+	if (!aes256_ecb_encrypt(key, tmp[0], output))
+		return 0;
+	return 1;
+}
+
+static int eoi_decrypt_can(struct sc_pkcs15_u8 *enc_can, char *can) {
+	/* Magic key that is used to decrypt CAN */
+	const u8 magic_key[32] = {0xC8, 0x12, 0x0F, 0xD8, 0x21, 0x20, 0x1F, 0x77, 0xF1, 0x83, 0x9D, 0xD8, 0x86, 0xB0, 0x5C, 0xF2, 0x4F, 0x7E, 0x52, 0x66, 0xE5, 0x87, 0x89, 0x2B, 0xF4, 0xC5, 0xE5, 0x4C, 0x54, 0xA1, 0x55, 0x30};
+	u8 can_key[32] = { 0 };
+
+	if (!can || !enc_can || !enc_can->value || enc_can->len != 24)
+		return SC_ERROR_INVALID_ARGUMENTS;
+
+	if (!get_can_key(magic_key, 0x01, enc_can->value, &can_key[0]))
+		return SC_ERROR_INTERNAL;
+	if (!get_can_key(magic_key, 0x02, enc_can->value, &can_key[AES_BLOCK_SIZE]))
+		return SC_ERROR_INTERNAL;
+
+	if (!aes256_ecb_decrypt(can_key, enc_can->value, (u8 *)can))
+		return SC_ERROR_INTERNAL;
+	can[AES_BLOCK_SIZE - 1] = 0;
+
+	return SC_SUCCESS;
+}
+
+static int eoi_sm_open(struct sc_card *card)
+{
+	struct eoi_privdata *privdata = (struct eoi_privdata *)card->drv_data;
+	struct establish_pace_channel_input pace_input;
+	struct establish_pace_channel_output pace_output;
+	struct sc_path path;
+
+	if (!privdata)
+		return SC_ERROR_INTERNAL;
+
+	if (!privdata->can[0]) {
+		/* If no CAN is specified in conf, try to decrypt it from enc_can file */
+		LOG_TEST_RET(card->ctx, eoi_decrypt_can(&privdata->enc_can, privdata->can),
+			"Cannot decrypt CAN");
```

Comment:
please, use temporary variable here. Its much easier to read than calling a function in the macro (which might even have some side effects as evaluating the argument several times -- i think it is not the case here though) and breaking the line unnecessarily during a function/macro.
```suggestion
		int rc = eoi_decrypt_can(&privdata->enc_can, privdata->can);
		LOG_TEST_RET(card->ctx, rc, "Cannot decrypt CAN");
```
this anti-pattern is used several times (not only in this function) so I would suggest the change also in the other places

Reply:
Done.